### PR TITLE
Compound production

### LINKED ISSUE #570: Photosynthesisers should affect environmental oxygen
Split from this issue: https://github.com/Revolutionary-Games/Thrive/issues/553

- [ ] It should be possible for the cell species to affect the amount of oxygen and CO2 (photosynthesisers)

### PR DESCRIPTION
**Pull request by Maxonovien**
**Brief Description of What This PR Does**

This PR adds production/consumption on gas compounds in patches depending on population in them.
It also paves the way for further dynamism in patches composition.

Planned changes:
- [x] Add compound amount manipulation infrastructure (patch volume...)
- [x] Have input/output of processes from microbes affect the world
- [x] Add a save upgrader
- [x] Add gas output to microbe processes (to avoid depletion)
- ~~[ ] Add patch volume in GUI (larger patches are more stable) & translation support~~
- [x] Add constant input to patches for balance
- [x] Balance production factor to get an interesting/not game-breaking evolution of patches

**Related Issues**

Closes #570.

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author.

### CODE REVIEW

File: src/microbe_stage/PatchMapGenerator.cs [hunk 1]
```diff
@@ -77,7 +77,7 @@ public static PatchMap Generate(WorldGenerationSettings settings, Species defaul
         map.AddPatch(bathypelagic);
 
         var abyssopelagic = new Patch("PATHCH_PANGONIAN_ABYSSOPELAGIC", 5,
-            GetBiomeTemplate("abyssopelagic"))
+            GetBiomeTemplate("abyssopelagic"), (long)(2000 * 2000) * 2000)
```
Comment (hhyyrylainen) [inline src/microbe_stage/PatchMapGenerator.cs:80]: `2000L`?

File: src/general/IWorldEffect.cs [hunk 2]
```diff
@@ -53,3 +54,71 @@ public void OnTimePassed(double elapsed, double totalTimePassed)
...
+        this.targetWorld = targetWorld;
+    }
+
+    public void OnRegisterToWorld()
+    {
+    }
+
+    public void OnTimePassed(double elapsed, double totalTimePassed)
+    {
+        List<Compound> gasCompounds = SimulationParameters.Instance.GetGasCompounds();
```
Comment (adQuid) [inline src/general/IWorldEffect.cs:74]: Will this list be changing? I'd look into doing this just once in OnRegisterToWorld

File: src/microbe_stage/Patch.cs [hunk 3]
```diff
@@ -23,6 +23,12 @@ public class Patch
     [JsonProperty]
     public readonly int[] Depth = new int[2] { -1, -1 };
 
+    /// <summary>
+    ///   The volume of the patch, in meters cube.
```
Comment (adQuid) [inline src/microbe_stage/Patch.cs:27]: Do we really want to restrict ourselves to exact units here?

File: src/microbe_stage/ProcessSpeedInformation.cs [hunk 4]
```diff
@@ -37,7 +37,11 @@ public ProcessSpeedInformation(BioProcess process)
     public IReadOnlyDictionary<Compound, float> FullSpeedRequiredEnvironmentalInputs =>
         WritableFullSpeedRequiredEnvironmentalInputs;
 
-    public IEnumerable<KeyValuePair<Compound, float>> Outputs => WritableOutputs;
+    public IEnumerable<KeyValuePair<Compound, float>> Outputs =>
+        WritableOutputs.Where(p => !p.Key.IsEnvironmental);
+
+    public IEnumerable<KeyValuePair<Compound, float>> EnvironmentalOutputs =>
+        WritableOutputs.Where(p => p.Key.IsEnvironmental);
```
Comment (adQuid) [inline src/microbe_stage/ProcessSpeedInformation.cs:44]: Why were these needed?

**Review by adQuid — COMMENTED** [review]
Overall surprisingly simple. No complaints on the grand scale.

File: src/general/IWorldEffect.cs (hunk 2 above)
Comment (hhyyrylainen) [reply src/general/IWorldEffect.cs:74]: I'd slightly prefer this being here, because after all this runs just once for each editor entry. And re-checking the gas compounds list each time would make this more compatible with loading mods after creating a save that add more compound types.

**Comment by hhyyrylainen** [comment]
Is this related to https://github.com/Revolutionary-Games/Thrive/issues/570 ?

File: simulation_parameters/SimulationParameters.cs [hunk 5]
```diff
@@ -184,6 +184,26 @@ public List<Compound> GetCloudCompounds()
         return result;
     }
 
+    /// <summary>
+    ///   Returns all *molecular* compounds that are dissolved in the environment, i.e. gas.
+    /// </summary>
+    /// <remarks>This excludes sunlight</remarks>
```
[the remark this answers is not in the document]
Comment (Maxonovien) [reply simulation_parameters/SimulationParameters.cs:190]: No, it is actually limited to environmental compounds like CO2, O2 and N. Perhaps some rephrasing could phrase this better indeed.

File: src/microbe_stage/Patch.cs (hunk 3 above)
Comment (Maxonovien) [reply src/microbe_stage/Patch.cs:27]: Tough question that has been haunting my nights for a while (do I exaggerate here? Yes, definitely.).

Depth itself is in meters, and so the volume, defined as deltaDepth^3 is in meters cube. 
I feel like having real units helps getting a clear view of what values represent physically, but I agree it is a restriction.

I don't have much of an opinion on this.

File: src/microbe_stage/ProcessSpeedInformation.cs (hunk 4 above)
Comment (Maxonovien) [reply src/microbe_stage/ProcessSpeedInformation.cs:44]: These are not useful as of now because there is no environmental output, but it will start to be useful when some processes like photosynthesis will start rejecting environmental compounds. Yet, the previous infrastructure assumes there is no environmental outputs, so I filtered them out to keep things working.

Having access to Environmental ones rather than the full list is just sugar to mimick inputs and help readability.

**Comment by Maxonovien** [comment]
It is. I mostly put the draft up to get it accessible as I wanted Thim's opinion on it before it gets any more complicated, so I did not bother to link issues. I'll have to check for other relations, but I can add this issue to the list already.

**Comment by buckly90** [comment]
You're making great progress, I'm happy with how it's turning out. Noticed oxygen became very abundant in the surface patches, while the depths were very abundant with CO2.

The most glaring problems I see so far are that patches end up with little to no CO2 or O2, depending on where you are; as well as the effects of gasses on processes making them become very powerful.
Oh, and the tidepool was stripped of it's gasses rather spontaneously, which was odd.

I suggest that the rate of gas exchange between patches be increased so that the oxygen and CO2 produced in both environments mixes a little better. In addition, and this doesn't need to be done now, I believe we should adjust how processes scale with gasses so that they grant diminishing returns.

Keep up the great work!

**Comment by Kasterisk** [comment]
Any numerical compound amount percentage display in the UI needs rounding to only 3 decimal places, similar to how the compound amounts is handled.

File: src/microbe_stage/PatchMapGenerator.cs [hunk 6]
```diff
@@ -16,7 +16,7 @@ public static PatchMap Generate(WorldGenerationSettings settings, Species defaul
 
         // Predefined patches
         var vents = new Patch("PATCH_PANGONIAN_VENTS", 0,
-            GetBiomeTemplate("aavolcanic_vent"))
+            GetBiomeTemplate("aavolcanic_vent"), 500 * 500 * 500)
```
Comment (Kasterisk) [inline src/microbe_stage/PatchMapGenerator.cs:19]: Can a `Math.Pow` be used for these?

File: src/saving/ISaveUpgradeStep.cs [hunk 7]
```diff
@@ -104,7 +105,7 @@ internal class UpgradeStep055To056 : BaseRecursiveJSONWalkerStep
         protected override void CheckAndUpdateProperty(JProperty property)
         {
             var children = property.Value.Children<JProperty>();
-            var childrenNames = children.Select(c => c.Name);
+            var childrenNames = children.Select(c => c.Name).ToList();
```
[the remark this answers is not in the document]
Comment (Maxonovien) [reply src/saving/ISaveUpgradeStep.cs:108]: Oh, it's probably an old change (I think I originally had the save upgrade step in this call), and I did not remove all the changes I did when moving all to a new step. I'll correct this tomorrow. Good catch!

File: simulation_parameters/SimulationParameters.cs [hunk 8]
```diff
@@ -190,6 +190,26 @@ public List<Compound> GetCloudCompounds()
         return result;
     }
 
+    /// <summary>
+    ///   Returns all environmental *molecular* compounds that are dissolved in the environment, i.e. gas.
+    /// </summary>
+    /// <remarks>This excludes sunlight, and includes O2, CO2, N...</remarks>
+    public List<Compound> GetGasCompounds()
```
[the remark this answers is not in the document]
Comment (Maxonovien) [reply simulation_parameters/SimulationParameters.cs:197]: Get Cloud compounds was similar to my implementation, but I agreed that it would be better as LINQ, so I changed both.

**Review by adQuid — CHANGES_REQUESTED** [review]
Just one thing I would change. I've also noticed that nothing releases nitrogen into the environment so we'll be dependent on the hand of God until we add something, but I think that that is outside the scope of this change.

File: src/general/IWorldEffect.cs [hunk 9]
```diff
@@ -93,3 +95,167 @@ public void OnTimePassed(double elapsed, double totalTimePassed)
...
+                        if (individualCompoundProduction.TryGetValue(compound, out CompoundBalance compoundProduction))
+                        {
+                            if (!compoundsProduced.ContainsKey(compound))
+                                compoundsProduced[compound] = 0;
+
+                            compoundsProduced[compound] += compoundProduction.Balance * patch.SpeciesInPatch[species] *
+                                Constants.DISSOLVED_PRODUCTION_FACTOR;
+                        }
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (Maxonovien) [reply src/general/IWorldEffect.cs:146]: Shouldn't that start with a non-capital letter? But good idea otherwise?

**Comment by Maxonovien** [comment]
Oh, you mean nothing releases it IRL? Actually, there might be some azote trapped within cavities getting slowly rejected to the ocean, or whatever.

File: src/general/IWorldEffect.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (adQuid) [reply src/general/IWorldEffect.cs:146]: You're right; that's what I get for coding in github.

**Comment by adQuid** [comment]
> Oh, you mean nothing releases it IRL? Actually, there might be some nitrogen trapped within cavities getting slowly rejected to the ocean, or whatever.

I have no idea how real science works. I just meant in the simulation no biological process releases it, so we have to remember to never turn off the magic nitrogen hose in a future update.

File: src/microbe_stage/Patch.cs [hunk 10]
```diff
@@ -24,6 +24,12 @@ public class Patch
...
```
Comment (hhyyrylainen) [inline src/microbe_stage/Patch.cs:28]:
```suggestion
    ///   The volume of the patch, in meters cubed.
```
Isn't "meters cube" a typo? At least I've seen more "cubed" so it sounds more correct to me.

File: src/microbe_stage/ProcessSpeedInformation.cs (hunk 4 above)
Comment (hhyyrylainen) [reply src/microbe_stage/ProcessSpeedInformation.cs:43]: I'd like a comment here detailing this aspect so that when someone is fixing this up properly they'll have the necessary context to know to change these back.

File: src/saving/ISaveUpgradeStep.cs [hunk 11]
```diff
@@ -162,6 +163,174 @@ private void UpgradeBehaviouralValues(JProperty property, JEnumerable<JProperty>
...
+            {
+                if (depth[1].Value<int?>() == null || depth[0].Value<int?>() == null)
+                {
+                    throw new ArgumentException("Depth array does not start with two integers!");
+                }
+
+                var depthDifference = depth[1].Value<int>() - depth[0].Value<int>();
+
+                // Assume cubic patches for upgrade
+                container.Add("Volume", (long)Math.Pow(depthDifference, 3));
```
Comment (hhyyrylainen) [inline src/saving/ISaveUpgradeStep.cs:268]: Why is this done like this? Surely it would be better to either: specify the volume in the biome definitions *or* also update the patch generator to use the same depth difference logic, otherwise I fear there's too much highly specific math duplication between this and the patch generation.

File: src/saving/ISaveUpgradeStep.cs [hunk 12]
```diff
@@ -162,6 +163,174 @@ private void UpgradeBehaviouralValues(JProperty property, JEnumerable<JProperty>
...
+        ///       ...
+        ///   </para>
+        /// </remarks>
+        private void UpgradePatchesVolume(JObject container, JEnumerable<JProperty> children)
+        {
+            if (children.First(p => p.Name == "Depth").Value is JArray depth)
+            {
+                if (depth[1].Value<int?>() == null || depth[0].Value<int?>() == null)
+                {
+                    throw new ArgumentException("Depth array does not start with two integers!");
```
Comment (hhyyrylainen) [inline src/saving/ISaveUpgradeStep.cs:262]: Should throw json exceptions like the other upgrade steps as these are about invalid json data.

File: src/saving/ISaveUpgradeStep.cs [hunk 13]
```diff
@@ -162,6 +163,174 @@ private void UpgradeBehaviouralValues(JProperty property, JEnumerable<JProperty>
...
+            var childrenNames = children.Select(c => c.Name).ToList();
+
+            // Add volume for named patches (properties)
+            if (childrenNames.Contains("Depth") && !childrenNames.Contains("Volume"))
+            {
+                UpgradePatchesVolume((JObject)property.Value, children);
+            }
+
+            // Most patches are defined through references to non-named tokens of array "Adjacent"
+            if (property.Name == "Adjacent")
```
Comment (hhyyrylainen) [inline src/saving/ISaveUpgradeStep.cs:182]: I'm a bit suspicious about this as I thought that the base step should already recurse through everything... so in my head, this design should not be required. It seems the call to the base CheckAndUpdateProperty is missing in this method.

File: simulation_parameters/SimulationParameters.cs [hunk 14]
```diff
@@ -177,17 +178,16 @@ public bool DoesCompoundExist(string name)
...
-            {
-                result.Add(entry.Value);
-            }
-        }
+        return compounds.Values.Where(c => c.IsCloud).ToList();
+    }
 
-        return result;
+    /// <summary>
+    ///   Returns all environmental *molecular* compounds that are dissolved in the environment, i.e. gas.
```
Comment (hhyyrylainen) [inline simulation_parameters/SimulationParameters.cs:185]: I feel like this comment should be a bit reworded to be just "Returns compounds that are dissolved in the environment (aren't clouds)" just so that the comment won't become incorrect in the future.

File: simulation_parameters/SimulationParameters.cs [hunk 15]
```diff
@@ -177,17 +178,16 @@ public bool DoesCompoundExist(string name)
...
+    /// </summary>
+    /// <remarks>This excludes sunlight, and includes O2, CO2, N...</remarks>
```
Comment (hhyyrylainen) [inline simulation_parameters/SimulationParameters.cs:187]: remarks needs to have a para tag in it and it needs to be formatted like shown in the styleguide.

File: simulation_parameters/SimulationParameters.cs [hunk 16]
```diff
@@ -177,17 +178,16 @@ public bool DoesCompoundExist(string name)
...
+    public List<Compound> GetGasCompounds()
+    {
+        return compounds.Values.Where(c => c.IsCloud && c.IsEnvironmental && c.CanBeDistributed).ToList();
```
Comment (hhyyrylainen) [inline simulation_parameters/SimulationParameters.cs:190]: I feel a bit confused by these flag checks here. It *should* be the case that IsCloud and IsEnvironmental can't be true for the same compound. Also CanBeDistributed is related to colony compound sharing, so this confuses me a lot.

File: src/general/IWorldEffect.cs [hunk 17]
```diff
@@ -93,3 +95,167 @@ public void OnTimePassed(double elapsed, double totalTimePassed)
...
+    {
+        List<Compound> gasCompounds = SimulationParameters.Instance.GetGasCompounds();
+
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            // First, add the constant intake to the patch, since it will influence the available amount for species.
+            // TODO move to own effect
+            AddConstantIntakeToPatch(patch);
+
+            Dictionary<Compound, float> compoundsProduced = new Dictionary<Compound, float>(gasCompounds.Count);
```
Comment (hhyyrylainen) [inline src/general/IWorldEffect.cs:124]: You seem to be a bit underusing `var` in general, but especially this line is unnecessarily long due to not using `var`.

File: src/general/IWorldEffect.cs [hunk 18]
```diff
@@ -93,3 +95,167 @@ public void OnTimePassed(double elapsed, double totalTimePassed)
...
+                }
+            }
+
+            compoundsProduced = ScaleByOverconsumption(compoundsProduced, patch);
+
+            foreach (var compound in compoundsProduced.Keys)
+            {
+                if (patch.Biome.Compounds.TryGetValue(compound, out EnvironmentalCompoundProperties compoundValue))
+                {
+                    // TODO if capped here, use something to scale production
```
Comment (hhyyrylainen) [inline src/general/IWorldEffect.cs:155]:
```suggestion
                    // TODO: if capped here, use something to scale production
```

File: src/general/IWorldEffect.cs [hunk 19]
```diff
@@ -93,3 +95,167 @@ public void OnTimePassed(double elapsed, double totalTimePassed)
...
+    public void AddConstantIntakeToPatch(Patch patch)
+    {
+        var oxygen = SimulationParameters.Instance.GetCompound("oxygen");
+        var carbonDioxide = SimulationParameters.Instance.GetCompound("carbondioxide");
+        var nitrogen = SimulationParameters.Instance.GetCompound("nitrogen");
+        var constantCompoundsDissolvedIntake = new Dictionary<Compound, float>(3);
+
+        constantCompoundsDissolvedIntake[oxygen] = Constants.PATCH_CONSTANT_OXYGEN_INPUT;
+        constantCompoundsDissolvedIntake[carbonDioxide] = Constants.PATCH_CONSTANT_CARBON_DIOXYDE_INPUT;
+        constantCompoundsDissolvedIntake[nitrogen] = Constants.PATCH_CONSTANT_NITROGEN_INPUT;
```
Comment (hhyyrylainen) [inline src/general/IWorldEffect.cs:181]: You can rewrite this as a dictionary with initializer.

File: src/general/IWorldEffect.cs [hunk 20]
```diff
@@ -93,3 +95,167 @@ public void OnTimePassed(double elapsed, double totalTimePassed)
...
+
+        foreach (var intake in constantCompoundsDissolvedIntake)
+        {
+            if (patch.Biome.Compounds.TryGetValue(intake.Key, out EnvironmentalCompoundProperties compoundValue))
+            {
+                // TODO if capped here, use something to scale production
```
Comment (hhyyrylainen) [inline src/general/IWorldEffect.cs:187]:
```suggestion
                // TODO: if capped here, use something to scale production
```

File: src/general/IWorldEffect.cs [hunk 21]
```diff
@@ -93,3 +95,167 @@ public void OnTimePassed(double elapsed, double totalTimePassed)
...
+    /// <summary>
+    ///   Reduce ALL compound production by a scaling factor if one compounds happens to be overproduced.
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     This is a heuristic to prevent production of compounds from non-existing ones.
+    ///     A better yet more complicated approach would be to only limit the processes consuming the target compounds.
+    ///   </para>
+    /// </remarks>
+    public Dictionary<Compound, float> ScaleByOverconsumption(Dictionary<Compound, float> compoundsProduced,
```
Comment (hhyyrylainen) [inline src/general/IWorldEffect.cs:246]: Shouldn't this return void as this modifies the parameter in-place?

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
I've now reviewed all the code here and left quite many code style comments. I didn't see any big issues regarding the code architecture.

File: src/saving/ISaveUpgradeStep.cs [hunk 22]
```diff
@@ -162,6 +163,174 @@ private void UpgradeBehaviouralValues(JProperty property, JEnumerable<JProperty>
...
+                }
+
+                string targetWorldRef = effectsArray[0]["targetWorld"]?["$ref"]?.ToString() ??
+                    throw new JsonException("Missing field from the first element of effect array: " +
+                        "targetWorld or its $ref!");
+
+                var typeTokenText = "'$type': '" + gasProductionEffectName + "'";
+                var targetWorldTokenText = "'targetWorld': { '$ref': '" + targetWorldRef + "' }";
+                var effectTokenText = "{" + typeTokenText + "," + targetWorldTokenText + "}";
+                var effectToken = JToken.Parse(effectTokenText);
```
Comment (hhyyrylainen) [inline src/saving/ISaveUpgradeStep.cs:322]: Wouldn't it be better to actually build the JObject properly instead of doing a bunch of string concatenation?
(I've seen some r/programminghorrorposts about manually created JSON strings...)

File: src/general/IWorldEffect.cs (hunk 21 above)
Comment (Maxonovien) [reply src/general/IWorldEffect.cs]: I think that for some reason it was not actually doing in place changes, so I used a return instead

File: src/general/IWorldEffect.cs [hunk 23]
```diff
@@ -93,3 +95,167 @@ public void OnTimePassed(double elapsed, double totalTimePassed)
...
+                    patch.Biome.Compounds[compound] = new EnvironmentalCompoundProperties
+                    {
+                        Amount = 0, Density = 0, Dissolved = compoundsProduced[compound] / patch.Volume,
+                    };
+                }
+            }
+        }
+    }
+
+    public void AddConstantIntakeToPatch(Patch patch)
```
[the remark this answers is not in the document]
Comment (Maxonovien) [reply src/general/IWorldEffect.cs]: Should, I also swapped a bunch of similar methods as such.

File: src/general/IWorldEffect.cs (hunk 21 above)
Comment (hhyyrylainen) [reply src/general/IWorldEffect.cs]: `struct` would be a value type, but Dictionary is a `class`

File: src/general/IWorldEffect.cs [hunk 24]
```diff
@@ -93,3 +95,167 @@ public void OnTimePassed(double elapsed, double totalTimePassed)
...
+                    compoundValue.Dissolved = Math.Max(
+                        compoundValue.Dissolved + compoundsProduced[compound] / patch.Volume,
+                        Constants.DISSOLVED_MIN);
+                    patch.Biome.Compounds[compound] = compoundValue;
+                }
+                else
+                {
+                    patch.Biome.Compounds[compound] = new EnvironmentalCompoundProperties
+                    {
+                        Amount = 0, Density = 0, Dissolved = compoundsProduced[compound] / patch.Volume,
```
[the remark this answers is not in the document]
Comment (Maxonovien) [reply src/general/IWorldEffect.cs]: I did a shared method that should use it properly.

File: src/general/IWorldEffect.cs (hunk 17 above)
Comment (Maxonovien) [reply src/general/IWorldEffect.cs]: I guess old habits die hard. Fixed.

File: simulation_parameters/SimulationParameters.cs (hunk 16 above)
Comment (Maxonovien) [reply simulation_parameters/SimulationParameters.cs:190]: I agree that by the JSON compounds, this should return an empty list, yet since changes are applied, it probably doesn't... I'm as confused as you on this, I don't recall why I used `IsCloud`... I'll have to test this when I'll have a moment.

`CanBeDistributed` is used to exclude sunlight. That was both the best heuristic I had at hand, and something logical, since globally "gas compounds" should be spread in colonies, unlike whatever compound type is sunlight.

Comment (hhyyrylainen) [reply simulation_parameters/SimulationParameters.cs:190]: Perhaps if there isn't a good existing flag for it you should add a new property to the Compound class that specifies that it is a gas?

File: src/saving/ISaveUpgradeStep.cs (hunk 11 above)
Comment (Maxonovien) [reply src/saving/ISaveUpgradeStep.cs:268]: Patch generator is hardcoded. I used cubes as a simple model for patches, but nothing should force patches to be cubes. Now this upgrade step just intends to give a volume to patches of older versions, and since I can't access the exact value (unless I just move every patch volume to `Constants.cs` but then any modification would require to change multiple files at once, which seems terrible an idea, if I may), I just go for the simple model of a cube, which is enough to get older versions to be *playable*.

The only way I see to change this would be to rework the patch generator to work from either Constants or JSONs, and that would be a PR of its own.

File: src/microbe_stage/Patch.cs (hunk 10 above)
Comment (Maxonovien) [reply src/microbe_stage/Patch.cs:28]: Actually, the official naming is "cubic metre", as it seems : https://en.wikipedia.org/wiki/Cubic_metre

I'll change that.

File: src/saving/ISaveUpgradeStep.cs (hunk 11 above)
Comment (hhyyrylainen) [reply src/saving/ISaveUpgradeStep.cs:268]: I more suggested that you add the "Volume" as a parameter for a biome, then use that as the value here and in the patch map generator. I guess if we don't care about the math being off (and thus perhaps balancing as well) when comparing upgraded and other saves, this is fine if that is explained with a comment why this approach is taken.

**Comment by Maxonovien** [comment]
I guess, I just keep forgetting that `out` creates variables outside of the `if` scope...

File: src/microbe_stage/Patch.cs (hunk 10 above)
Comment (hhyyrylainen) [reply src/microbe_stage/Patch.cs:28]: At least don't change the spelling to "metre" that spelling is disgusting to me as it's farther away from how the word is actually said (as a Finnish person I'm always offended when a word is not spelled the same way it is pronounced).

**Comment by hhyyrylainen** [comment]
> I guess, I just keep forgetting that out creates variables outside of the if scope...

Yeah, that's something that I didn't know initially either, but I've now found that it is helpful in reducing the amount of code that is needed.

**Comment by buckly90** [comment]
It has been some time since I last looked at this PR and now it seems that the compounds barely change at all now.
I tested this PR with two separate playthroughs, one as a strict carnivorous hunter, and another as a docile photosynthesizer; Both concluded with the same results.

I'm unsure if this is because cells no longer have a significant impact, or if I just happened to live in a world that struck a perfect balance both times.

Either way, when I have some time on my hands I'll see about playing with the numbers and see if I can make things feel right for testing purposes.

File: src/microbe_stage/Patch.cs (hunk 10 above)
Comment (Maxonovien) [reply src/microbe_stage/Patch.cs:28]: Well, the thing is, the guideline is to keep by British English, whose spelling is actually "metre". I'd go against adding exception for one's sole personal taste (especially given that we'd then have to change about half the vocabulary by this metric --English be English :) )

File: src/saving/ISaveUpgradeStep.cs (hunk 11 above)
Comment (Maxonovien) [reply src/saving/ISaveUpgradeStep.cs:268]: You mean having a predetermined volume for a biome? I see no reason why we should have a predefined volume for caves : I think this goes along the same logic as the one used for depth measures.  I'll go along the comment route, I think.

Ultimately, I guess this shows how much the current patch generator system should be updated...

File: simulation_parameters/SimulationParameters.cs (hunk 16 above)
Comment (Maxonovien) [reply simulation_parameters/SimulationParameters.cs:190]: I guess it would make sense to have a good flag for sunlight (and potential other similar compounds), but I'm short for ideas at the moment. Unless someone comes by with a great idea, I'll simply add a "isSunlight" flag whenever I have time to wrap everything up.

**Comment by Maxonovien** [comment]
In which patch are you playing ? Remember that larger patch are less affected by the same amount of cells (and since we have very tiny patches like the pond, it must be a bit controlled). 

Also, obviously the idea is not to get it to change drastically every generation, otherwise it would be fairly unplayable. But sure, some help in balancing this would be great!

**Comment by buckly90** [comment]
> In which patch are you playing ? Remember that larger patch are less affected by the same amount of cells (and since we have very tiny patches like the pond, it must be a bit controlled).

I tried all of the surface patches, and the vents.

The vents had no visible change at all, but the epipelagic and tidepools atleast had an occasional 0.0001% or so change in concentration

**Comment by Maxonovien** [comment]
Maybe I went to harsh on the species then ? Well, I'll give it a look when I have time. But as far as I remember I had some decent changes in some places... Perhaps is it also due to a small pop? I'll see.

File: src/saving/ISaveUpgradeStep.cs (hunk 11 above)
Comment (hhyyrylainen) [reply src/saving/ISaveUpgradeStep.cs:268]: I was suggesting that as it would fit the current patch generation model and also make it consistent when upgrading saves.
But feel free to do this:
> I guess if we don't care about the math being off (and thus perhaps balancing as well) when comparing upgraded and other saves, this is fine if that is explained with a comment why this approach is taken.

File: simulation_parameters/SimulationParameters.cs (hunk 16 above)
Comment (hhyyrylainen) [reply simulation_parameters/SimulationParameters.cs:194]: Why not "IsGas"?

File: src/microbe_stage/Patch.cs (hunk 10 above)
Comment (hhyyrylainen) [reply src/microbe_stage/Patch.cs:28]: https://github.com/Revolutionary-Games/Thrive/pull/3107

**Review by Kasterisk — CHANGES_REQUESTED** [review]

File: src/general/GameWorld.cs [hunk 25]
```diff
@@ -72,6 +72,9 @@ public GameWorld()
 
         // Register glucose reduction
         TimedEffects.RegisterEffect("reduce_glucose", new GlucoseReductionEffect(this));
+
+        // Register gas compounds production
+        TimedEffects.RegisterEffect("produce_gases", new GasProductionEffect(this));
```
[the remark this answers is not in the document]
Comment (hhyyrylainen) [reply src/general/GameWorld.cs:77]: Actually seems like this is correct form as this is the plural of gas and not a verb.
https://grammarist.com/usage/gases-gasses/

**Review by adQuid — APPROVED** [review]

**Comment by Maxonovien** [comment]
Currently impeded due to the fact that thylakoid species still somewhat thrive without photosynthesis once they appeared. We thus need photosynthetic elements to be a burden under low CO2/high O2. See photosynthesis.

File: src/general/IWorldEffect.cs [hunk 26]
```diff
@@ -93,3 +95,160 @@ public void OnTimePassed(double elapsed, double totalTimePassed)
...
+        this.targetWorld = targetWorld;
+    }
+
+    public void OnRegisterToWorld()
+    {
+    }
+
+    public void OnTimePassed(double elapsed, double totalTimePassed)
+    {
+        List<Compound> gasCompounds = SimulationParameters.Instance.GetGasCompounds();
```
Comment (Kemikal1) [inline src/general/IWorldEffect.cs:116]: i debugged this and for some reason this always gets me gasCompunds.Count = 0

File: simulation_parameters/SimulationParameters.cs (hunk 16 above)
Comment (Kemikal1) [reply simulation_parameters/SimulationParameters.cs:194]: For some  reason those flags are never simultaneously true?

**Comment by hhyyrylainen** [comment]
I won't close this PR as there's still probably a chance that this will be finished.

**Comment by hhyyrylainen** [comment]
I'll close this PR as a pretty old one that is quite unlikely to resume work anytime soon so the upcoming history rewrite will cause this PR to basically conflict with everything and need manual fixing.
